src/parser.rs

```diff
@@ -1271,3 +1289,114 @@ fn definition_location(
     definition.range().start,
   )
 }
+
+struct SymbolVisibility {
+  root_exported_ids: HashMap<UniqueSymbolId, IndexSet<UniqueSymbolId>>,
+  /// Symbol identifiers that are not exported, but are referenced
+  /// by exported symbols.
+  non_exported_public_ids: HashSet<UniqueSymbolId>,
+}
+
+impl SymbolVisibility {
+  pub fn build(
+    graph: &ModuleGraph,
+    root_symbol: &deno_graph::type_tracer::RootSymbol,
+  ) -> Result<Self, DocError> {
```

Comment:
I realized while working on this that the "type tracing" feature in deno_graph is somewhat useless and no longer used. We should refactor it to a general purpose symbol graph: https://github.com/denoland/deno_graph/issues/316

Reply:
Agreed